### PR TITLE
RA-1516 Added support to select the location from the userProperty in the Login Screen

### PR DESCRIPTION
**Pull request by suthagar23**
# Description 

This PR contains the changes to select the session location from userProperty in the login screen. It will be worked according to these following steps,

1. It will check for the Global property(GP) called "referenceapplication.locationUserPropertyName". 
2. If the GP has the value, then it will hide the location selector from the login screen
3. When users try to log in, it will fetch the location from the user property
4. If the respected user doesn't have the required user property, then it will redirect to the usual login screen(with login selector) to select the location

image

image

# Ticket 

Ticket : https://issues.openmrs.org/browse/RA-1516

### CODE REVIEW

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java [hunk 1]
```diff
@@ -206,24 +225,46 @@ public String post(@RequestParam(value = "username", required = false) String us
...
+						}
+                    }
+				}
+			}
+
 
-			try {
-				Context.authenticate(username, password);
 
+			//TODO uncomment this to replace the if clause after it
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: When do you plan to work on the above TODO?

Comment (suthagar23) [reply omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: Actually this was came from the exiting code. I just moved that whole part. Do you want me to clear this as well?

Comment (suthagar23) [reply omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: Just removed @dkayiwa

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java [hunk 2]
```diff
@@ -78,11 +81,12 @@ public String overrideLoginpage() {
 	 * @should set the referer as the redirectUrl in the page model if referer URL is within context path
 	 */
 	public String get(PageModel model,
-	                  UiUtils ui,
-	                  PageRequest pageRequest,
-	                  @CookieValue(value = COOKIE_NAME_LAST_SESSION_LOCATION, required = false) String lastSessionLocationId,
-	                  @SpringBean("locationService") LocationService locationService,
-	                  @SpringBean("appFrameworkService") AppFrameworkService appFrameworkService) {
+					  UiUtils ui,
+					  PageRequest pageRequest,
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: Did you intentionally change this indention?

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java [hunk 3]
```diff
@@ -110,11 +114,24 @@ public String get(PageModel model,
 			Context.removeProxyPrivilege(GET_LOCATIONS);
 		}
 
+        model.addAttribute("showSessionLocations", !isLocationUserPropertyAvailable(administrationService));
+		Object showLocation = pageRequest.getAttribute("showSessionLocations");
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: The indention here does not look right

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java [hunk 4]
```diff
@@ -110,11 +114,24 @@ public String get(PageModel model,
 			Context.removeProxyPrivilege(GET_LOCATIONS);
 		}
 
+        model.addAttribute("showSessionLocations", !isLocationUserPropertyAvailable(administrationService));
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: Did you intentionally duplicate the model.addAttribute call?

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java [hunk 5]
```diff
@@ -110,11 +114,24 @@ public String get(PageModel model,
...
+		if(showLocation != null && showLocation.toString().equals("true")) {
+            model.addAttribute("showSessionLocations", true);
+        }
 		model.addAttribute("lastSessionLocation", lastSessionLocation);
 
 		return null;
 	}
 
+	private boolean isLocationUserPropertyAvailable(AdministrationService administrationService) {
+		String showLocationSelector = administrationService.getGlobalProperty(ReferenceApplicationConstants.LOCATION_USER_PROPERTY_NAME);
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: What if it is available but does not have a value of true?

File: omod/src/main/webapp/pages/login.gsp [hunk 6]
```diff
@@ -19,67 +19,73 @@
...
-      			jQuery('#login-form').submit();
-    		}
-		});
-		var  listItem = Array.from(jQuery('#sessionLocation li'));
-		for (var i in  listItem){
-			 listItem[i].setAttribute('data-key', i);
-			 listItem[i].addEventListener('keyup', function (event){
-				var keyCode = event.which || event.keyCode;
-				switch (keyCode) {
-					case 37: // move left
```
Comment (dkayiwa) [inline omod/src/main/webapp/pages/login.gsp]: Too many indention changes. Can we do or fix the indention using another pull request? It gets hard to see what actually changed.

**Comment by dkayiwa** [comment]
We also prefer to attach screenshot on the ticket instead of pull request as per https://wiki.openmrs.org/display/docs/Pull+Request+Tips

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java (hunk 4 above)
Comment (suthagar23) [reply omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: Nope @dkayiwa, It used to set that value always, and if wanted it will update this value by next line

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java (hunk 5 above)
Comment (suthagar23) [reply omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: Actually, I decided to use only one GP for this task, So if that GP contains any value except "false", it will return that value as the userProperty name

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java [hunk 7]
```diff
@@ -241,34 +281,33 @@ public String post(@RequestParam(value = "username", required = false) String us
 						//don't redirect back to the login page on success nor an external url
 						if (isUrlWithinOpenmrs(pageRequest, redirectUrl)) {
 							if (!redirectUrl.contains("login.") && isSameUser(pageRequest, username)) {
-                                if (log.isDebugEnabled())
-                                    log.debug("Redirecting user to " + redirectUrl);
-                                return "redirect:" + redirectUrl;
-                            } else {
-                                if (log.isDebugEnabled())
-                                    log.debug("Redirect contains 'login.', redirecting to home page");
```
[the remark this answers is not in the document]
Comment (suthagar23) [reply omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: Done @dkayiwa

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java [hunk 8]
```diff
@@ -277,7 +316,8 @@ public String post(@RequestParam(value = "username", required = false) String us
 		//TODO limit login attempts by IP Address
 
 		pageRequest.getSession().setAttribute(SESSION_ATTRIBUTE_REDIRECT_URL, redirectUrl);
-
+		// Since the user is already authenticated without location, need to logout before redirecting
+		Context.logout();
```
[the remark this answers is not in the document]
Comment (suthagar23) [reply omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: Since I need to do the context.authenticate() before verifying the location restrictions. So if anything required isn't satisfied, then it will be redirected with the authenticated user. So the user should automatically log out from the system before the redirect.

File: omod/src/main/webapp/pages/login.gsp (hunk 6 above)
Comment (suthagar23) [reply omod/src/main/webapp/pages/login.gsp]: I have done the refactor for those indentions. It occured automatically since I done it through the IntelliJ. Could be better to go through this :smile:

File: omod/src/main/webapp/pages/login.gsp [hunk 9]
```diff
@@ -19,67 +19,74 @@
...
-    	updateSelectedOption = function() {
-	        jQuery('#sessionLocation li').removeClass('selected');
-	        
-			var sessionLocationVal = jQuery('#sessionLocationInput').val();
-	        if(sessionLocationVal != null && sessionLocationVal != "" && sessionLocationVal != 0){
-	            jQuery('#sessionLocation li[value|=' + sessionLocationVal + ']').addClass('selected');
-	        }
-    	};
-    
+        updateSelectedOption = function() {
```
Comment (suthagar23) [inline omod/src/main/webapp/pages/login.gsp]: This indentation should be fixed since wrong in the original file - https://github.com/openmrs/openmrs-module-referenceapplication/blob/master/omod/src/main/webapp/pages/login.gsp#L25

File: omod/src/main/webapp/pages/login.gsp [hunk 10]
```diff
@@ -19,67 +19,74 @@
...
+        };
+
         updateSelectedOption();
 
         jQuery('#sessionLocation li').click( function() {
             jQuery('#sessionLocationInput').val(jQuery(this).attr("value"));
             updateSelectedOption();
         });
-		jQuery('#sessionLocation li').focus( function() {
+        jQuery('#sessionLocation li').focus( function() {
```
Comment (suthagar23) [inline omod/src/main/webapp/pages/login.gsp]: This indentation should be fixed since wrong in the original file -  https://github.com/openmrs/openmrs-module-referenceapplication/blob/master/omod/src/main/webapp/pages/login.gsp#L39

**Comment by suthagar23** [comment]
@dkayiwa  I have attached the affected login screen to the ticket comment. Please have a look again.

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java (hunk 5 above)
Comment (dkayiwa) [reply omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: You have a gp named "referenceapplication.locationUserPropertyName" and then expect a value of true or false? This does not look intuitive because from the look of the gp name, i would expect it to contain the actual name.

File: omod/src/main/webapp/pages/login.gsp (hunk 9 above)
Comment (dkayiwa) [reply omod/src/main/webapp/pages/login.gsp]: If you are fixing indention, then i would recommend a separate pull request for it. That way, it becomes easier for me to tell the difference between indention changes and those require to implement this functionality.

Comment (suthagar23) [reply omod/src/main/webapp/pages/login.gsp]: So then, I will remove those from here as well

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java (hunk 5 above)
Comment (suthagar23) [reply omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: We decided to use one GP for this, So when evet it is set, we know that it true, if not set or has value as "false" we know it is false.

Comment (dkayiwa) [reply omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: What happens when it has a value of a non existing property?

File: omod/src/main/webapp/pages/login.gsp (hunk 6 above)
Comment (suthagar23) [reply omod/src/main/webapp/pages/login.gsp]: I have updated all the indentions changes

Comment (dkayiwa) [reply omod/src/main/webapp/pages/login.gsp]: Isn't this another indention?

Comment (suthagar23) [reply omod/src/main/webapp/pages/login.gsp]: Extremely sorry @dkayiwa  It happened automatically with my IDE. I just made it manually now. Please have a look again

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java [hunk 11]
```diff
@@ -110,11 +114,24 @@ public String get(PageModel model,
 			Context.removeProxyPrivilege(GET_LOCATIONS);
 		}
 
+		model.addAttribute("showSessionLocations", !isLocationUserPropertyAvailable(administrationService));
+		Object showLocation = pageRequest.getAttribute("showSessionLocations");
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: I still do not understand why you call model.addAttribute twice instead of once

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java [hunk 12]
```diff
@@ -110,11 +114,24 @@ public String get(PageModel model,
...
+		if(showLocation != null && showLocation.toString().equals("true")) {
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: Are you sure we are expecting this value to be true?

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java [hunk 13]
```diff
@@ -110,11 +114,24 @@ public String get(PageModel model,
...
+		    model.addAttribute("showSessionLocations", true);
+		}
 		model.addAttribute("lastSessionLocation", lastSessionLocation);
 
 		return null;
 	}
 
+	private boolean isLocationUserPropertyAvailable(AdministrationService administrationService) {
+		String locationUserPropertyName = administrationService.getGlobalProperty(ReferenceApplicationConstants.LOCATION_USER_PROPERTY_NAME);
+		if(StringUtils.isNotBlank(locationUserPropertyName) && !locationUserPropertyName.equals("false")) {
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: I thought we agreed to simply use a valid user property name. Instead instead of true or false

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java (hunk 12 above)
Comment (suthagar23) [reply omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: Yes, Initial login URL will be "http://localhost:8080/openmrs/login.htm", and then if the user doesn't have the location userproperty, then it will redirect to "http://localhost:8080/openmrs/referenceapplication/login.page?showSessionLocations=true&"

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java (hunk 11 above)
Comment (dkayiwa) [reply omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: Can't you do it once?

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java (hunk 13 above)
Comment (suthagar23) [reply omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: Yes , Whenever it is set, we know that it is true. If not set, we know it is false.

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java [hunk 14]
```diff
@@ -253,22 +291,21 @@ public String post(@RequestParam(value = "username", required = false) String us
...
-				    ui.message(ReferenceApplicationConstants.MODULE_ID + ".error.login.fail"));
+						ui.message("referenceapplication.login.error.locationRequired"));
+			} else {
+				// the UI shouldn't allow this, but protect against it just in case
+				pageRequest.getSession().setAttribute(ReferenceApplicationWebConstants.SESSION_ATTRIBUTE_ERROR_MESSAGE,
+						ui.message("referenceapplication.login.error.invalidLocation", sessionLocation.getName()));
 			}
+		}
+		catch (ContextAuthenticationException ex) {
+			if (log.isDebugEnabled())
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: Did you take a look at this? https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-CodeStyle

Comment (suthagar23) [reply omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: Actually, this came from the existing code. Do you want me to remove this as well?

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java (hunk 11 above)
Comment (suthagar23) [reply omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: Done @dkayiwa

File: api/src/main/java/org/openmrs/module/referenceapplication/ReferenceApplicationConstants.java [hunk 15]
```diff
@@ -23,4 +23,6 @@
 	
 	public static final long PROCESS_HL7_TASK_INTERVAL = 5L;
 
+	public static final String LOCATION_USER_PROPERTY_NAME = "referenceapplication.locationUserPropertyName";
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/module/referenceapplication/ReferenceApplicationConstants.java]: After looking at the other pull request, i have noticed that you expect the user property to be named as locationUuid

In such a case, of what value is this global property?

Comment (suthagar23) [reply api/src/main/java/org/openmrs/module/referenceapplication/ReferenceApplicationConstants.java]: Using this Global property(name - referenceapplication.locationUserPropertyName), I want to fetch the userProperty Name(should be value of this GP). Then need to use this userPropertyName to fetch the location from the userProperty(should be value of that userProperty).

So in our module, I'm using "locationUuuid" as the userPropertyName to store the locationUuid. So setting "locationUuid" to this GP.

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/module/referenceapplication/ReferenceApplicationConstants.java]: If you know that the property will be named as "locationUuuid", then of what value is the gp?

Comment (suthagar23) [reply api/src/main/java/org/openmrs/module/referenceapplication/ReferenceApplicationConstants.java]: So can we move the locationUserProperty to a global property name as "locationbasedaccess.locationUserPropertyName"? So the userProperty name also will be fetched from the GP, and use that same value to this RefApp GP as well?

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/module/referenceapplication/ReferenceApplicationConstants.java]: What do you think? What are the pros and cons of either side?

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java [hunk 16]
```diff
@@ -110,11 +114,26 @@ public String get(PageModel model,
 			Context.removeProxyPrivilege(GET_LOCATIONS);
 		}
 
+		Boolean isLocationUserPropertyAvailable = isLocationUserPropertyAvailable(administrationService);
+		Object showLocation = pageRequest.getAttribute("showSessionLocations");
+		if(showLocation != null && showLocation.toString().equals("true")) {
+			// if the request contains a attribute as showSessionLocations, then ignore isLocationUserPropertyAvailable
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: Confising how showLocation is true and then you set is location user property available to false?

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java [hunk 17]
```diff
@@ -206,24 +227,43 @@ public String post(@RequestParam(value = "username", required = false) String us
...
-			cookie.setHttpOnly(true);
-			pageRequest.getResponse().addCookie(cookie);
+		try {
+			Context.authenticate(username, password);
+			String locationUserPropertyName = administrationService.getGlobalProperty(ReferenceApplicationConstants.LOCATION_USER_PROPERTY_NAME);
+			if (StringUtils.isNotBlank(locationUserPropertyName)) {
+				if (Context.isAuthenticated() && Context.getUserContext().getAuthenticatedUser() != null) {
+					String locationUuid = Context.getUserContext().getAuthenticatedUser().getUserProperty(locationUserPropertyName);
+					if (StringUtils.isNotBlank(locationUuid)) {
+						sessionLocation = locationService.getLocationByUuid(locationUuid);
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: What happens when the uuid is for a non existing location?

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java (hunk 16 above)
Comment (suthagar23) [reply omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: if the URL has the parameter as "showSessionLocations" with value True, then we need to ignore the userProperty method(if exist or not). So simply setting isLocationUserPropertyAvailable=false to disable this feature to view the location selector in the login screen. (See the address bar of the browser in the attached images in the PR)

Comment (dkayiwa) [reply omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: who sets the showSessionLocations parameter to true and when?

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java (hunk 17 above)
Comment (suthagar23) [reply omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: Just added a fix to this problem. Please look at this line again @dkayiwa

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java (hunk 16 above)
Comment (suthagar23) [reply omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: When a user doesn't get the location userProperty, then it will redirect to the login page with showSessionLocations=true. So I'm checking this parameter to show/hide the location selector here.  See this line - https://github.com/openmrs/openmrs-module-referenceapplication/pull/47/files#diff-bd998e9f6ec90d297ceba7fb14c2fe66R249

File: api/src/main/java/org/openmrs/module/referenceapplication/ReferenceApplicationConstants.java (hunk 15 above)
Comment (suthagar23) [reply api/src/main/java/org/openmrs/module/referenceapplication/ReferenceApplicationConstants.java]: I think, We have already concluded this discussion :smile:

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java [hunk 18]
```diff
@@ -206,24 +227,43 @@ public String post(@RequestParam(value = "username", required = false) String us
...
+					}
+					if (sessionLocation != null) {
+                        sessionLocationId = sessionLocation.getLocationId();
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: Can you check the indention here?

Comment (suthagar23) [reply omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: Fixed @dkayiwa

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java [hunk 19]
```diff
@@ -206,24 +227,43 @@ public String post(@RequestParam(value = "username", required = false) String us
...
+				if (Context.isAuthenticated() && Context.getUserContext().getAuthenticatedUser() != null) {
+					String locationUuid = Context.getUserContext().getAuthenticatedUser().getUserProperty(locationUserPropertyName);
+					if (StringUtils.isNotBlank(locationUuid)) {
+						sessionLocation = locationService.getLocationByUuid(locationUuid);
+					}
+					if (sessionLocation != null) {
+						sessionLocationId = sessionLocation.getLocationId();
+                    }
+					else {
+					    pageRequest.getSession().setAttribute(ReferenceApplicationWebConstants.SESSION_ATTRIBUTE_ERROR_MESSAGE,
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: Can you look into the indention here?

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java [hunk 20]
```diff
@@ -206,24 +227,43 @@ public String post(@RequestParam(value = "username", required = false) String us
...
+			if (StringUtils.isNotBlank(locationUserPropertyName)) {
+				if (Context.isAuthenticated() && Context.getUserContext().getAuthenticatedUser() != null) {
+					String locationUuid = Context.getUserContext().getAuthenticatedUser().getUserProperty(locationUserPropertyName);
+					if (StringUtils.isNotBlank(locationUuid)) {
+						sessionLocation = locationService.getLocationByUuid(locationUuid);
+					}
+					if (sessionLocation != null) {
+						sessionLocationId = sessionLocation.getLocationId();
+                    }
+					else {
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: The formatting or braces around here do not look correct.

Comment (suthagar23) [reply omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: Fixed

File: omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java (hunk 19 above)
Comment (suthagar23) [reply omod/src/main/java/org/openmrs/module/referenceapplication/page/controller/LoginPageController.java]: Fixed
